File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/PlayerEventForwarder.swift
//
//  PlayerEventForwarder.swift
//  flutter_theoplayer_sdk_ios
//
//  Created by Hovig on 18/10/2023.
//

import Foundation
import THEOplayerSDK
import Flutter

class PlayerEventForwarder {
    
    private let theoplayer: THEOplayer
    private let flutterAPI: THEOplayerFlutterAPI
    
    private var sourceChangeEventListener: EventListener?
    private var playEventListener: EventListener?
    private var playingEventListener: EventListener?
    private var pauseEventListener: EventListener?
    private var waitingEventListener: EventListener?
    private var durationChangeEventListener: EventListener?
    private var progressEventListener: EventListener?
    private var timeUpdateEventListener: EventListener?
    private var rateChangeEventListener: EventListener?
    private var seekingEventListener: EventListener?
    private var seekedEventListener: EventListener?
    private var volumeChangeEventListener: EventListener?
    private var resizeEventListener: EventListener?
    private var endedEventListener: EventListener?
    private var errorEventListener: EventListener?
    private var destroyEventListener: EventListener?
    private var readyStateChangeEventListener: EventListener?
    private var loadStartEventListener: EventListener?
    private var loadedMetadataEventListener: EventListener?
    private var loadedDataEventListener: EventListener?
    private var canPlayEventListener: EventListener?
    private var canPlayThroughEventListener: EventListener?
    
    private let emptyCompletion: (Result<Void, FlutterError>) -> Void = {result in }
    
    init(theoplayer: THEOplayer, flutterAPI: THEOplayerFlutterAPI) {
        self.theoplayer = theoplayer
        self.flutterAPI = flutterAPI
    }
    
    func attachListeners() {
        sourceChangeEventListener = theoplayer.addEventListener(type: PlayerEventTypes.SOURCE_CHANGE, listener: {event in
            self.flutterAPI.onSourceChange(source: SourceTransformer.toFlutterSourceDescription(source: event.source), completion: self.emptyCompletion)
        })
        
        playEventListener = theoplayer.addEventListener(type: PlayerEventTypes.PLAY, listener: {event in
            self.flutterAPI.onPlay(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        playingEventListener = theoplayer.addEventListener(type: PlayerEventTypes.PLAYING, listener: {event in
            self.flutterAPI.onPlaying(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        pauseEventListener = theoplayer.addEventListener(type: PlayerEventTypes.PAUSE, listener: {event in
            self.flutterAPI.onPause(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        waitingEventListener = theoplayer.addEventListener(type: PlayerEventTypes.WAITING, listener: {event in
            self.flutterAPI.onWaiting(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        durationChangeEventListener = theoplayer.addEventListener(type: PlayerEventTypes.DURATION_CHANGE, listener: {event in
            self.flutterAPI.onDurationChange(duration: event.duration ?? 0, completion: self.emptyCompletion)
        })
        
        progressEventListener = theoplayer.addEventListener(type: PlayerEventTypes.PROGRESS, listener: {event in
            self.flutterAPI.onProgress(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        timeUpdateEventListener = theoplayer.addEventListener(type: PlayerEventTypes.TIME_UPDATE, listener: {event in
            var currentProgramDateTime: Int64?;
            if let timeIntervalSince1970 = event.currentProgramDateTime?.timeIntervalSince1970 {
                currentProgramDateTime = Int64(timeIntervalSince1970)
            }
            
            self.flutterAPI.onTimeUpdate(currentTime: event.currentTime, currentProgramDateTime: currentProgramDateTime, completion: self.emptyCompletion)
        })
        
        rateChangeEventListener = theoplayer.addEventListener(type: PlayerEventTypes.RATE_CHANGE, listener: {event in
            self.flutterAPI.onRateChange(currentTime: event.currentTime, playbackRate: event.playbackRate, completion: self.emptyCompletion)
        })
        
        seekingEventListener = theoplayer.addEventListener(type: PlayerEventTypes.SEEKING, listener: {event in
            self.flutterAPI.onSeeking(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        seekedEventListener = theoplayer.addEventListener(type: PlayerEventTypes.SEEKED, listener: {event in
            self.flutterAPI.onSeeked(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        volumeChangeEventListener = theoplayer.addEventListener(type: PlayerEventTypes.VOLUME_CHANGE, listener: {event in
            self.flutterAPI.onVolumeChange(currentTime: event.currentTime, volume: Double(event.volume), completion: self.emptyCompletion)
        })
        
        resizeEventListener = theoplayer.addEventListener(type: PlayerEventTypes.RESIZE, listener: {event in
            self.flutterAPI.onResize(
                currentTime: self.theoplayer.currentTime,
                width: Int64(self.theoplayer.videoWidth),
                height: Int64(self.theoplayer.videoHeight),
                completion: self.emptyCompletion)
        })
        
        endedEventListener = theoplayer.addEventListener(type: PlayerEventTypes.ENDED, listener: {event in
            self.flutterAPI.onEnded(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        errorEventListener = theoplayer.addEventListener(type: PlayerEventTypes.ERROR, listener: {event in
            self.flutterAPI.onError(error: event.error, completion: self.emptyCompletion)
        })
        
        destroyEventListener = theoplayer.addEventListener(type: PlayerEventTypes.DESTROY, listener: {event in
            self.flutterAPI.onDestroy(completion: self.emptyCompletion)
        })
        
        readyStateChangeEventListener = theoplayer.addEventListener(type: PlayerEventTypes.READY_STATE_CHANGE, listener: {event in
            self.flutterAPI.onReadyStateChange(currentTime: event.currentTime, readyState: PlayerEnumTransformer.toFlutterReadyState(readyState: event.readyState), completion: self.emptyCompletion)
        })
        
        loadStartEventListener = theoplayer.addEventListener(type: PlayerEventTypes.LOAD_START, listener: {event in
            self.flutterAPI.onLoadStart(completion: self.emptyCompletion)
        })
        
        loadedMetadataEventListener = theoplayer.addEventListener(type: PlayerEventTypes.LOADED_META_DATA, listener: {event in
            self.flutterAPI.onLoadedMetadata(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        loadedDataEventListener = theoplayer.addEventListener(type: PlayerEventTypes.LOADED_DATA, listener: {event in
            self.flutterAPI.onLoadedData(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        canPlayEventListener = theoplayer.addEventListener(type: PlayerEventTypes.CAN_PLAY, listener: {event in
            self.flutterAPI.onCanPlay(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
        
        canPlayThroughEventListener = theoplayer.addEventListener(type: PlayerEventTypes.CAN_PLAY_THROUGH, listener: {event in
            self.flutterAPI.onCanPlayThrough(currentTime: event.currentTime, completion: self.emptyCompletion)
        })
    }
    
    // TODO: remove force unwraps
    func removeListeners() {
        theoplayer.removeEventListener(type: PlayerEventTypes.SOURCE_CHANGE, listener: sourceChangeEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.PLAY, listener: playEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.PLAYING, listener: playingEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.PAUSE, listener: pauseEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.WAITING, listener: waitingEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.DURATION_CHANGE, listener: durationChangeEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.PROGRESS, listener: progressEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.TIME_UPDATE, listener: timeUpdateEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.RATE_CHANGE, listener: rateChangeEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.SEEKING, listener: seekingEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.SEEKED, listener: seekedEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.VOLUME_CHANGE, listener: volumeChangeEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.RESIZE, listener: resizeEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.ENDED, listener: endedEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.ERROR, listener: errorEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.DESTROY, listener: destroyEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.READY_STATE_CHANGE, listener: readyStateChangeEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.LOAD_START, listener: loadStartEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.LOADED_META_DATA, listener: loadedMetadataEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.LOADED_DATA, listener: loadedDataEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.CAN_PLAY, listener: canPlayEventListener!)
        theoplayer.removeEventListener(type: PlayerEventTypes.CAN_PLAY_THROUGH, listener: canPlayThroughEventListener!)
    }
    
}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/flutter/PlayerEventForwarder.kt
package com.theoplayer.flutter

import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.player.CanPlayEvent
import com.theoplayer.android.api.event.player.CanPlayThroughEvent
import com.theoplayer.android.api.event.player.DestroyEvent
import com.theoplayer.android.api.event.player.DurationChangeEvent
import com.theoplayer.android.api.event.player.EndedEvent
import com.theoplayer.android.api.event.player.ErrorEvent
import com.theoplayer.android.api.event.player.LoadStartEvent
import com.theoplayer.android.api.event.player.LoadedDataEvent
import com.theoplayer.android.api.event.player.LoadedMetadataEvent
import com.theoplayer.android.api.event.player.PauseEvent
import com.theoplayer.android.api.event.player.PlayEvent
import com.theoplayer.android.api.event.player.PlayerEventTypes
import com.theoplayer.android.api.event.player.PlayingEvent
import com.theoplayer.android.api.event.player.ProgressEvent
import com.theoplayer.android.api.event.player.RateChangeEvent
import com.theoplayer.android.api.event.player.ReadyStateChangeEvent
import com.theoplayer.android.api.event.player.ResizeEvent
import com.theoplayer.android.api.event.player.SeekedEvent
import com.theoplayer.android.api.event.player.SeekingEvent
import com.theoplayer.android.api.event.player.SourceChangeEvent
import com.theoplayer.android.api.event.player.TimeUpdateEvent
import com.theoplayer.android.api.event.player.VolumeChangeEvent
import com.theoplayer.android.api.event.player.WaitingEvent
import com.theoplayer.android.api.player.Player
import com.theoplayer.flutter.pigeon.THEOplayerFlutterAPI
import com.theoplayer.flutter.transformers.PlayerEnumTransformer
import com.theoplayer.flutter.transformers.SourceTransformer

class PlayerEventForwarder(
    private val player: Player,
    private val flutterAPI: THEOplayerFlutterAPI
) {
    private val emptyCallback: (Result<Unit>) -> Unit = {}

    private val sourceChangeEventListener = EventListener<SourceChangeEvent> {
        flutterAPI.onSourceChange(SourceTransformer.toFlutterSourceDescription(it.source), emptyCallback)
    }

    private val playEventListener = EventListener<PlayEvent> {
        flutterAPI.onPlay(it.currentTime, emptyCallback)
    }

    private val playingEventListener = EventListener<PlayingEvent> {
        flutterAPI.onPlaying(it.currentTime, emptyCallback)
    }

    private val pauseEventListener = EventListener<PauseEvent> {
        flutterAPI.onPause(it.currentTime, emptyCallback)
    }

    private val waitingEventListener = EventListener<WaitingEvent> {
        flutterAPI.onWaiting(it.currentTime.toDouble(), emptyCallback)
    }

    private val durationChangeEventListener = EventListener<DurationChangeEvent> {
        flutterAPI.onDurationChange(it.duration, emptyCallback)
    }

    private val progressEventListener = EventListener<ProgressEvent> {
        flutterAPI.onProgress(it.currentTime, emptyCallback)
    }

    private val timeUpdateEventListener = EventListener<TimeUpdateEvent> {
        flutterAPI.onTimeUpdate(it.currentTime, it.currentProgramDateTime?.time, emptyCallback)
    }

    private val rateChangeEventListener = EventListener<RateChangeEvent> {
        flutterAPI.onRateChange(it.currentTime, it.playbackRate, emptyCallback)
    }

    private val seekingEventListener = EventListener<SeekingEvent> {
        flutterAPI.onSeeking(it.currentTime, emptyCallback)
    }

    private val seekedEventListener = EventListener<SeekedEvent> {
        flutterAPI.onSeeked(it.currentTime, emptyCallback)
    }

    private val volumeChangeEventListener = EventListener<VolumeChangeEvent> {
        flutterAPI.onVolumeChange(it.currentTime, it.volume, emptyCallback)
    }

    private val resizeEventListener = EventListener<ResizeEvent> {
        flutterAPI.onResize(it.currentTime, it.width.toLong(), it.height.toLong(), emptyCallback)
    }

    private val endedEventListener = EventListener<EndedEvent> {
        flutterAPI.onEnded(it.currentTime, emptyCallback)
    }

    private val errorEventListener = EventListener<ErrorEvent> {
        flutterAPI.onError(it.errorObject.toString(), emptyCallback)
    }

    private val destroyEventListener = EventListener<DestroyEvent> {
        flutterAPI.onDestroy(emptyCallback)
    }

    private val readyStateChangeEventListener = EventListener<ReadyStateChangeEvent> {
        flutterAPI.onReadyStateChange(it.currentTime, PlayerEnumTransformer.toFlutterReadyState(it.readyState), emptyCallback)
    }

    private val loadStartEventListener = EventListener<LoadStartEvent> {
        flutterAPI.onLoadStart(emptyCallback)
    }

    private val loadedMetadataEventListener = EventListener<LoadedMetadataEvent> {
        flutterAPI.onLoadedMetadata(it.currentTime, emptyCallback)
    }

    private val loadedDataEventListener = EventListener<LoadedDataEvent> {
        flutterAPI.onLoadedData(it.currentTime.toDouble(), emptyCallback)
    }

    private val canPlayEventListener = EventListener<CanPlayEvent> {
        flutterAPI.onCanPlay(it.currentTime, emptyCallback)
    }

    private val canPlayThroughEventListener = EventListener<CanPlayThroughEvent> {
        flutterAPI.onCanPlayThrough(it.currentTime, emptyCallback)
    }

    fun attachListeners() {
        player.addEventListener(PlayerEventTypes.SOURCECHANGE, sourceChangeEventListener)
        player.addEventListener(PlayerEventTypes.PLAY, playEventListener)
        player.addEventListener(PlayerEventTypes.PLAYING, playingEventListener)
        player.addEventListener(PlayerEventTypes.PAUSE, pauseEventListener)
        player.addEventListener(PlayerEventTypes.WAITING, waitingEventListener)
        player.addEventListener(PlayerEventTypes.DURATIONCHANGE, durationChangeEventListener)
        player.addEventListener(PlayerEventTypes.PROGRESS, progressEventListener)
        player.addEventListener(PlayerEventTypes.TIMEUPDATE, timeUpdateEventListener)
        player.addEventListener(PlayerEventTypes.RATECHANGE, rateChangeEventListener)
        player.addEventListener(PlayerEventTypes.SEEKING, seekingEventListener)
        player.addEventListener(PlayerEventTypes.SEEKED, seekedEventListener)
        player.addEventListener(PlayerEventTypes.VOLUMECHANGE, volumeChangeEventListener)
        player.addEventListener(PlayerEventTypes.RESIZE, resizeEventListener)
        player.addEventListener(PlayerEventTypes.ENDED, endedEventListener)
        player.addEventListener(PlayerEventTypes.ERROR, errorEventListener)
        player.addEventListener(PlayerEventTypes.DESTROY, destroyEventListener)
        player.addEventListener(PlayerEventTypes.READYSTATECHANGE, readyStateChangeEventListener)
        player.addEventListener(PlayerEventTypes.LOADSTART, loadStartEventListener)
        player.addEventListener(PlayerEventTypes.LOADEDMETADATA, loadedMetadataEventListener)
        player.addEventListener(PlayerEventTypes.LOADEDDATA, loadedDataEventListener)
        player.addEventListener(PlayerEventTypes.CANPLAY, canPlayEventListener)
        player.addEventListener(PlayerEventTypes.CANPLAYTHROUGH, canPlayThroughEventListener)
    }

    fun removeListeners() {
        player.removeEventListener(PlayerEventTypes.SOURCECHANGE, sourceChangeEventListener)
        player.removeEventListener(PlayerEventTypes.PLAY, playEventListener)
        player.removeEventListener(PlayerEventTypes.PLAYING, playingEventListener)
        player.removeEventListener(PlayerEventTypes.PAUSE, pauseEventListener)
        player.removeEventListener(PlayerEventTypes.WAITING, waitingEventListener)
        player.removeEventListener(PlayerEventTypes.DURATIONCHANGE, durationChangeEventListener)
        player.removeEventListener(PlayerEventTypes.PROGRESS, progressEventListener)
        player.removeEventListener(PlayerEventTypes.TIMEUPDATE, timeUpdateEventListener)
        player.removeEventListener(PlayerEventTypes.RATECHANGE, rateChangeEventListener)
        player.removeEventListener(PlayerEventTypes.SEEKING, seekingEventListener)
        player.removeEventListener(PlayerEventTypes.SEEKED, seekedEventListener)
        player.removeEventListener(PlayerEventTypes.VOLUMECHANGE, volumeChangeEventListener)
        player.removeEventListener(PlayerEventTypes.RESIZE, resizeEventListener)
        player.removeEventListener(PlayerEventTypes.ENDED, endedEventListener)
        player.removeEventListener(PlayerEventTypes.ERROR, errorEventListener)
        player.removeEventListener(PlayerEventTypes.DESTROY, destroyEventListener)
        player.removeEventListener(PlayerEventTypes.READYSTATECHANGE, readyStateChangeEventListener)
        player.removeEventListener(PlayerEventTypes.LOADSTART, loadStartEventListener)
        player.removeEventListener(PlayerEventTypes.LOADEDMETADATA, loadedMetadataEventListener)
        player.removeEventListener(PlayerEventTypes.LOADEDDATA, loadedDataEventListener)
        player.removeEventListener(PlayerEventTypes.CANPLAY, canPlayEventListener)
        player.removeEventListener(PlayerEventTypes.CANPLAYTHROUGH, canPlayThroughEventListener)
    }
}